### PR TITLE
[IMP]  functions: added some functions

### PR DESCRIPTION
**Pull request by matho-odoo**
## Description:

Co-authored-by: Pierre Rousseau <pro-odoo@users.noreply.github.com>

Task: [5231138](https://www.odoo.com/odoo/2328/tasks/5231138)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/functions/module_lookup.test.ts [hunk 1]
```diff
@@ -1922,3 +1922,30 @@ describe("OFFSET formula", () => {
...
+  test("should truncate decimal index before using it", () => {
+    expect(evaluateCell("A1", { A1: `=CHOOSE(1.4, "A", "B", "C")` })).toBe("A");
+  });
+  test("should convert string index to number", () => {
+    expect(evaluateCell("A1", { A1: `=CHOOSE("1", "A", "B", "C")` })).toBe("A");
+  });
+  test("error if index is not a number", () => {
+    expect(evaluateCell("A1", { A1: `=CHOOSE("a", "A", "B", "C")` })).toBe("#ERROR");
+  });
+  test("error if index is négative", () => {
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("error if index is negative", () => {
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 2]
```diff
@@ -1104,3 +1104,23 @@ export const OFFSET = {
...
+    arg("index (number)", _t("Which choice (of the up to 30 provided) to return.")),
+    arg(
+      "choice (ANY, RANGE<ANY>, repeating)",
+      _t(
+        "A potential value to return. Required. May be a reference to a cell or an individual value.."
+      )
+    ),
+  ],
+  compute: function (index: Maybe<FunctionResultObject>, ...choices: Arg[]) {
+    return choices[Math.floor(toNumber(index, this.locale)) - 1] || "#ERROR";
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: You can provide an error message to guide the user and help them understand what's wrong.

```suggestion
    const _index = Math.floor(toNumber(index, this.locale)) - 1;
    if (_index < 0 || _index >= choices.length) {
      return new EvaluationError(_("Index for CHOOSE is invalid. Valid values are between 1 and %s inclusive.", choices.length));
    }
    return choices[_index];
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 3]
```diff
@@ -1104,3 +1104,23 @@ export const OFFSET = {
...
+
+// -----------------------------------------------------------------------------
+// CHOOSE
+// -----------------------------------------------------------------------------
+export const CHOOSE = {
+  description: _t("An element from a list of choices based on index."),
+  args: [
+    arg("index (number)", _t("Which choice (of the up to 30 provided) to return.")),
+    arg(
+      "choice (ANY, RANGE<ANY>, repeating)",
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: nitpicking: to be consistent with all existing argument definitions, keep it lower-case
```suggestion
      "choice (any, range<any>, repeating)",
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 4]
```diff
@@ -1104,3 +1104,23 @@ export const OFFSET = {
...
   },
 } satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
+// CHOOSE
+// -----------------------------------------------------------------------------
+export const CHOOSE = {
+  description: _t("An element from a list of choices based on index."),
+  args: [
+    arg("index (number)", _t("Which choice (of the up to 30 provided) to return.")),
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: we don't have this limitation
```suggestion
    arg("index (number)", _t("Which choice to return.")),
```

File: tests/functions/module_lookup.test.ts [hunk 5]
```diff
@@ -1922,3 +1922,30 @@ describe("OFFSET formula", () => {
...
+    expect(evaluateCell("A1", { A1: `=CHOOSE(-1, "A", "B", "C")` })).toBe("#ERROR");
+  });
+  test("error if index is too big", () => {
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts:1945]: you can also test the error message
```ts
  test("error if index is too big", () => {
    const model = createModelFromGrid({ A1: `=CHOOSE(4, "A", "B", "C")` })
    expect(getEvaluatedCell(model, "A1").value).toBe("#ERROR");
    expect(getEvaluatedCell(model, "A1").message).toBe("--The error message--");
  });
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 6]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+  compute: function (
+    array: Matrix<{ value: string }>,
+    rows: Maybe<FunctionResultObject>,
+    columns: Maybe<FunctionResultObject>
+  ) {
+    const _rows = toNumber(rows, this.locale);
+    const _columns = toNumber(columns, this.locale);
+    let result = array;
+    if (Math.abs(_columns) >= array.length || Math.abs(_rows) >= array[0].length) {
+      return new EvaluationError(_t("Number of rows or column to exclude is too big."));
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: this is the error, but the message is not helpful. When creating user facing errors, we should try to be as helpful as possible so the user can correct his error.
Could be "The number of rows or column to exclude must be smaller or equal than the number of elements in the array"

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 7]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+
+// -----------------------------------------------------------------------------
+// ARRAYTOTEXT
+// -----------------------------------------------------------------------------
+
+export const ARRAYTOTEXT = {
+  description: _t(
+    "returns an array of text values from any specified range. It passes text values unchanged, and converts non-text values to text."
+  ),
+  //c'est ce que dit excel mais je trouve pas ça très clair
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: to be removed

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 8]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+  ],
+  compute: function (
+    array: Matrix<{ value: string }>,
+    format: Maybe<FunctionResultObject> = { value: 0 }
+  ) {
+    let result = "";
+    const symb = { 0: ",", 1: ";" };
+    const _format = toNumber(format, this.locale);
+    if (!(_format in Object.keys(symb))) {
+      return new EvaluationError(_t("Format must be 0 or 1"));
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: missing a test that has the wrong format

File: tests/functions/module_lookup.test.ts [hunk 9]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    };
+    //prettier-ignore
+    const result = {
+      A1: "A1", B1: "B1", 
+      A2: "A2", B2: "B2", 
+                                    D5: "A1", E5: "B1",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("error if too much rows to exclude", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("error if too many rows to exclude", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 10]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    const result = {
+      A1: "A1",
+      B1: "B1",
+      A2: "A2",
+      B2: "B2",
+      D5: "#ERROR",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("error if too much columns to exclude", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("error if too many columns to exclude", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 11]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    const result = {
+      A1: "A1",
+      B1: "B1",
+      A2: "A2",
+      B2: "B2",
+      D5: "#ERROR",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("error if too much rows to exclude in negative", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("error if too many rows to exclude in negative", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 12]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    //prettier-ignore
+    const grid = {
+      A1: "A1", B1: "B1",
+      A2: "A2", B2: "B2",
+                                    D5: "=DROP(A1:B2,-2)",
+    };
+    expect(evaluateCell("D5", grid)).toBe("#ERROR");
+  });
+  test("error if too much columns to exclude in negative", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("error if too many columns to exclude in negative", () => {
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 13]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+          result = result + ",";
+        }
+      }
+      if (row !== array[0].length - 1) {
+        result = result + symb[_format];
+      }
+    }
+    if (_format === 1) {
+      result = "{" + result + "}";
+    }
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: I have some doubts about the performance of the string concatenation solution.

File: tests/functions/module_lookup.test.ts [hunk 14]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+      A2: "A2", B2: "B2",
+                                    D5: "=DROP(A1:B2,0,-2)",
+    };
+    expect(evaluateCell("D5", grid)).toBe("#ERROR");
+  });
+});
+
+describe("ARRAYTOTEXT formula", () => {
+  test("correct without format", () => {
+    //prettier-ignore
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]: missing a bunch of tests that will convert things to text, from the top of my head: 
- links
- booleans
- numbers with decimals
- errors

File: tests/functions/module_lookup.test.ts [hunk 15]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    };
+    //prettier-ignore
+    const result = {
+      A1: "A1", B1: "B1", 
+      A2: "A2", B2: "B2", 
+                                    D5: "A2", E5: "B2",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("all rows if too much rows to take", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("all rows if too many rows to take", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 16]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+    //prettier-ignore
+    const result = {
+      A1: "A1", B1: "B1", 
+      A2: "A2", B2: "B2", 
+                                    D5: "A1", E5: "B1",
+                                    D6: "A2", E6: "B2",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("all columns if too much columns to take", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("all columns if too many columns to take", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 17]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+      A2: "A2",
+      B2: "B2",
+      D5: "A1",
+      E5: "B1",
+      D6: "A2",
+      E6: "B2",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("all rows if too much rows to take in negative", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("all rows if too many rows to take in negative", () => {
```

File: tests/functions/module_lookup.test.ts [hunk 18]
```diff
@@ -1922,3 +1922,293 @@ describe("OFFSET formula", () => {
...
+      A2: "A2",
+      B2: "B2",
+      D5: "A1",
+      E5: "B1",
+      D6: "A2",
+      E6: "B2",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
+  });
+  test("all columns if too much columns to take in negative", () => {
```
Comment (VincentSchippefilt) [inline tests/functions/module_lookup.test.ts]:
```suggestion
  test("all columns if too many columns to take in negative", () => {
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 19]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+    return result;
+  },
+  isExported: true,
+} satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
+// FORMULATEXT
+// -----------------------------------------------------------------------------
+
+export const FORMULATEXT = {
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: missing the tests for this new formula

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 20]
```diff
@@ -1104,3 +1104,194 @@ export const OFFSET = {
...
+  compute: function (cellReference: { value: string }) {
+    const { sheetName, xc } = splitReference(cellReference.value);
+    const { col, row } = toCartesian(xc);
+    let sheetId: UID;
+    if (sheetName) {
+      //@ts-ignore
+      sheetId = this.getters.getSheetIdByName(sheetName);
+    } else {
+      sheetId = this.getters.getActiveSheetId();
+    }
```
Comment (VincentSchippefilt) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]:
```suggestion
const sheetId: UID = (sheetName && this.getters.getSheetIdByName(sheetName)) ?? this.getters.getActiveSheetId();
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts (hunk 13 above)
Comment (matho-odoo) [reply packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: I tried with a large formula dataset and execution time was acceptable

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 21]
```diff
@@ -1104,3 +1104,193 @@ export const OFFSET = {
     );
   },
 } satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
+// CHOOSE
+// -----------------------------------------------------------------------------
+export const CHOOSE = {
+  description: _t("An element from a list of choices based on index."),
```
Comment (hokolomopo) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: Probably missing a word in the description. `Chose an element` ? `Return an element` ?

File: tests/functions/module_lookup.test.ts [hunk 22]
```diff
@@ -1922,3 +1922,344 @@ describe("OFFSET formula", () => {
...
+      A2: "A2", B2: "B2",
+                                    D5: "=TAKE(A1:B2,-1)", E5: "", 
+    };
+    //prettier-ignore
+    const result = {
+      A1: "A1", B1: "B1", 
+      A2: "A2", B2: "B2", 
+                                    D5: "A2", E5: "B2",
+    };
+    expect(evaluateGrid(grid)).toEqual(result);
```
Comment (hokolomopo) [inline tests/functions/module_lookup.test.ts]:
```suggestion
    const expectedResult = {D5: "A2", E5: "B2",};
    expect(evaluateGrid(grid)).toMatchObject(expectedResult);
```

nitpick, I'd test only the relevant cells, we don't really need to test that A1:B2 is unchanged, it pollutes the test a bit.  Same comment for the other tests

File: tests/functions/module_lookup.test.ts [hunk 23]
```diff
@@ -1922,3 +1922,344 @@ describe("OFFSET formula", () => {
...
+  });
+  test("all rows if too many rows to take", () => {
+    //prettier-ignore
+    const grid = {
+      A1: "A1", B1: "B1",
+      A2: "A2", B2: "B2",
+                                    D5: "=TAKE(A1:B2,3)", E5: "",
+                                    D6: "", E6: ""
```
Comment (hokolomopo) [inline tests/functions/module_lookup.test.ts]: nitpick: defining empty cells is kinda useless here

File: package-lock.json [hunk 24]
```diff
@@ -126,6 +126,7 @@
       "integrity": "sha512-2BCOP7TN8M+gVDj7/ht3hsaO/B/n5oDbiAyyvnRlNOs+u1o+JWNYTQrmpuNp1/Wq2gcFrI01JAW+paEKDMx/CA==",
       "dev": true,
       "license": "MIT",
+      "peer": true,
```
Comment (LucasLefevre) [inline package-lock.json]: revert this.
We really should investigate why you always have those changes. Probably the version of npm or node

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 25]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+// CHOOSE
+// -----------------------------------------------------------------------------
+export const CHOOSE = {
+  description: _t("Returns an element from a list of choices based on index."),
+  args: [
+    arg("index (number)", _t("Which choice to return.")),
+    arg(
+      "choice (any, range<any>, repeating)",
+      _t(
+        "A potential value to return. Required. May be a reference to a cell or an individual value."
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]:
```suggestion
        "A potential value to return. May be a reference to a cell or an individual value."
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 26]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+    for (let row = 0; row < array[0].length; row++) {
+      for (let col = 0; col < array.length; col++) {
+        const value = array[col][row].value;
+        let text: string;
+        if (isEvaluationError(value)) {
+          text = value;
+        } else if (typeof value === "string") {
+          text = _format ? `"${value}"` : value;
+        } else {
+          text = toString(value);
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: This exact logic is already implemented in `evaluationResultToDisplayString`.
You can extract the code into a helper function and re-use it here

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 27]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+  args: [
+    arg("array (range)", _t("The array to convert into text")),
+    arg("format (number, default=0)", _t("The format of the returned data."), FORMAT_OPTIONS),
+  ],
+  compute: function (
+    array: Matrix<{ value: string }>,
+    format: Maybe<FunctionResultObject> = { value: 0 }
+  ) {
+    let result = "";
+    const symb = { 0: ",", 1: ";" };
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: It's not as simple as that. This depends on the locale (english, french, etc.) 😉 
But luckily, it's already implemented in `evaluationResultToDisplayString`

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 28]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+// -----------------------------------------------------------------------------
+
+export const FORMULATEXT = {
+  description: _t("Returns a formula as a string."),
+  args: [arg("cell_reference (meta)", _t("A reference to a cell."))],
+  compute: function (cellReference: { value: string }) {
+    const { sheetName, xc } = splitReference(cellReference.value);
+    const { col, row } = toCartesian(xc);
+    const sheetId: UID =
+      (sheetName && this.getters.getSheetIdByName(sheetName)) ?? this.getters.getActiveSheetId();
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: It's not correct to fallback on the active sheet because the formula could be on another sheet.
The meta argument is always provided fully qualified (with the sheet name), so there's no actual issue, but still, let be more correct
```suggestion
      (sheetName && this.getters.getSheetIdByName(sheetName)) ?? this.__originSheetId;
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 29]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+    const result = this.getters.getCell({ sheetId, col, row });
+    if (result && isFormula(result.content)) {
+      return result.content;
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: use more precise names where you can
```suggestion
    const cell = this.getters.getCell({ sheetId, col, row });
    if (cell && isFormula(cell.content)) {
      return cell.content;
```

File: demo/data.js [hunk 30]
```diff
@@ -1820,6 +1830,11 @@ export const demoData = {
         C226: "General Kenobi",
         C227: "Hello there",
         C228: "Kenobi",
+        C229: "B",
+        C230: "Kenobi",
+        C231: "{TEXTBEFORE,Hello there;REGEXEXTRACT,Kenobi}",
```
Comment (LucasLefevre) [inline demo/data.js]: shouldn't it be:
```suggestion
        C231: '{"TEXTBEFORE","Hello there";"REGEXEXTRACT","Kenobi"}',
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 31]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+// -----------------------------------------------------------------------------
+
+export const TAKE = {
+  description: _t(
+    "Returns a specified number of contiguous rows or columns from the start or end of an array."
+  ),
+  args: [
+    arg("array (range)", _t("The array from which to take rows or columns.")),
+    arg(
+      "rows (number)",
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: it is optional, right ?
```suggestion
      "rows (number, optional)",
```

File: tests/functions/module_lookup.test.ts [hunk 32]
```diff
@@ -1922,3 +1922,296 @@ describe("OFFSET formula", () => {
...
+    expect(evaluateCell("D5", grid)).toBe("#N/A");
+  });
+});
+describe("TAKE formula", () => {
+  test("correct for one row", () => {
+    //prettier-ignore
+    const grid = {
+      A1: "A1", B1: "B1",
+      A2: "A2", B2: "B2",
+                                    D5: "=TAKE(A1:B2,1)", E5: "",
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts]: There's a trailing space at the end of the line here.
You can install an extension to spot them.
```suggestion
                                    D5: "=TAKE(A1:B2,1)", E5: "",
```
Here is what it looks like in my IDE with [Trailing Spaces](https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces) installed

Image

File: tests/functions/module_lookup.test.ts [hunk 33]
```diff
@@ -1922,3 +1922,296 @@ describe("OFFSET formula", () => {
...
+    //prettier-ignore
+    const grid = {
+      A1: "", B1: "B1",
+      A2: "A2", B2: "B2",
+                                    D5: "=FORMULATEXT(A1)",
+    };
+    expect(evaluateCell("D5", grid)).toBe("#N/A");
+  });
+});
+describe("TAKE formula", () => {
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts:2070]: You can also add a test for the edge case where everything should be taken: `=TAKE(A1:B2,,)

File: tests/functions/module_lookup.test.ts [hunk 34]
```diff
@@ -1922,3 +1922,296 @@ describe("OFFSET formula", () => {
...
+    const expectedResult = { D5: "A2", E5: "B2" };
+    expect(evaluateGrid(grid)).toMatchObject(expectedResult);
+  });
+  test("all rows if too many rows to take", () => {
+    //prettier-ignore
+    const grid = {
+      A1: "A1", B1: "B1",
+      A2: "A2", B2: "B2",
+                                    D5: "=TAKE(A1:B2,3)", E5: "",
+                                    D6: "", E6: ""
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts]: I think the usual way to check the result when it's an array is to use `getRangeValuesAsMatrix` instead of defining empty cells.
You can ask Alexis

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 35]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+    return result;
+  },
+  isExported: true,
+} satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
+// ARRAYTOTEXT
+// -----------------------------------------------------------------------------
+
+export const ARRAYTOTEXT = {
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: ARRAYTOTEXT should be in the array category, in `module_array.ts`

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 36]
```diff
@@ -1104,3 +1114,207 @@ export const OFFSET = {
...
+    }
+  },
+  isExported: true,
+} satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
+// TAKE
+// -----------------------------------------------------------------------------
+
+export const TAKE = {
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts:1194]: I would move TAKE and DROP next to each other as they are related

File: tests/functions/module_lookup.test.ts [hunk 37]
```diff
@@ -1922,3 +1922,296 @@ describe("OFFSET formula", () => {
...
+      A2: "A2", B2: "B2",
+                                    D5: "=ARRAYTOTEXT(A1:B2)",
+    };
+    expect(evaluateCell("D5", grid)).toBe("A1,,A2,B2");
+  });
+  test("correct with links, booleans, decimals and errors", () => {
+    //prettier-ignore
+    const grid = {
+      A1: "www.odoo.com", B1: "true", C1: "12/12/2012",
+      A2: "3.14", B2: "#ERROR", C2: "",
```
Comment (LucasLefevre) [inline tests/functions/module_lookup.test.ts]:
```suggestion
      A2: "3.14", B2: "=0/0", C2: "",
```

File: packages/o-spreadsheet-engine/src/functions/module_lookup.ts [hunk 38]
```diff
@@ -1104,3 +1110,155 @@ export const OFFSET = {
     );
   },
 } satisfies AddFunctionDescription;
+
+// -----------------------------------------------------------------------------
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_lookup.ts]: nitpicking: those lines are a bit too long. They don't fit on one line in github or smaller screens

Image

File: packages/o-spreadsheet-engine/src/functions/module_array.ts [hunk 39]
```diff
@@ -14,83 +15,17 @@ import {
...
-      for (let row = 0; row < matrix[col].length; row++) {
-        result[col][currentRow + row] = matrix[col][row];
-      }
-    }
-    currentRow += matrix[0]?.length ?? 0;
-  }
-
-  return result;
-}
+const FORMAT_OPTIONS = [
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/functions/module_array.ts]: Move this close to where it's used. It's used only once

File: packages/o-spreadsheet-engine/src/helpers/matrix.ts [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+      .map((row) =>
+        row.map((val) => cellValueToDisplayString(val, emptyCell, locale)).join(rowSeparator)
+      )
+      .join(";");
+    return `{${arrayStr}}`;
+  }
+  return cellValueToDisplayString(result, emptyCell, locale);
+}
+
+export function cellValueToDisplayString(
```
Comment (LucasLefevre) [inline packages/o-spreadsheet-engine/src/helpers/matrix.ts]: doesn't need to be exported
```suggestion
function cellValueToDisplayString(
```
